### PR TITLE
Smarter fetching of 3rd party CNI binaries

### PR DESCRIPTION
**Pull request by heschlie**
Added target for fetching CNI binaries, and marker to avoid re-download when they exist.

### CODE REVIEW

**Comment by tomdee** [comment]
host-local is still being downloaded in two places with two versions - https://github.com/heschlie/calico-cni/blob/b08a4bcc822fdc722223f039b035eae764607800/Makefile#L191

**Comment by tomdee** [comment]
I'm still a little unsure that yo're using the best release for CNI - I think if you check what kubernetes uses you'll find they use a non official release

**Comment by tomdee** [comment]
I would prefer it if we avoided using an additional marker file for this. The existence of the plugins under `dist/` already tells us that they've been downloaded. What's your justification for needing a marker?

**Comment by heschlie** [comment]
I does look like kubernetes/hyperkube is using their own build

```
ARCH?=amd64
TEMP_DIR:=$(shell mktemp -d -t hyperkubeXXXXXX)
CNI_RELEASE=07a8a28637e97b22eb8dfe710eeae1344f69d16e
...
https://storage.googleapis.com/kubernetes-release/network-plugins/cni-${ARCH}-${CNI_RELEASE}.tar.gz
```

it points to a particular PR which was accepted after the latest release, we could just pull using their link I suppose?

As to my reasoning for using a marker is mostly my ignorance on how to do stuff with Makefiles, I'll do some digging and make it mo'betta.

Do we need to keep that host-local at the version it has?  Right now it is pulling v0.2.2 but the one we might want to pull for hyperkube uses a PR post v0.3.0 unless I'm mistaken (very possible)

**Comment by tomdee** [comment]
The existing host-local for the tests can be changed, the version we use for that is _much_ less important than the version we ship onto people's systems when they use our install scripts.

**Comment by caseydavenport** [comment]
@tomdee @heschlie what's the status of this? Is it being worked on or should it be closed?

**Comment by tomdee** [comment]
It's with @heschlie still.

**Comment by heschlie** [comment]
@tomdee thanks for the help this morning, finally some more changes for your review!

File: Makefile [hunk 1]
```diff
@@ -102,14 +103,15 @@ $(DEPLOY_CONTAINER_MARKER): Dockerfile build-containerized fetch-cni-bins
 	touch $@
 
 .PHONY: fetch-cni-bins
-fetch-cni-bins:
+fetch-cni-bins: dist/flannel dist/loopback dist/host-local
+
+dist/flannel dist/loopback dist/host-local:
 	mkdir -p dist
 	mkdir -p tmp-cni
```
Comment (tomdee) [inline Makefile]: I don't think you need this temp dir. Just extract the files to the right place

`tar -xzf cni-amd64-v0.4.0.tgz -C dist ./flannel ./loopback ./host-local` (but using curl of course)

File: Makefile [hunk 2]
```diff
@@ -33,7 +34,7 @@ docker-image: $(DEPLOY_CONTAINER_MARKER)
 
 .PHONY: clean
 clean:
-	rm -rf dist vendor $(DEPLOY_CONTAINER_MARKER) .go-pkg-cache
+	rm -rf dist vendor $(DEPLOY_CONTAINER_MARKER) .go-pkg-cache tmp-cni
```
Comment (tomdee) [inline Makefile]: Can avoid tmp-cni (see below)

File: Makefile [hunk 3]
```diff
@@ -18,6 +18,7 @@ MAKE_SURE_DIST_EXIST := $(shell mkdir -p dist)
 CALICO_BUILD?=calico/go-build
 DEPLOY_CONTAINER_NAME=calico/cni
 DEPLOY_CONTAINER_MARKER=cni_deploy_container.created
+FETCH_CNI_BINS_MARKER=fetch_cni_bins.created
```
Comment (tomdee) [inline Makefile]: Are you using this marker?

File: Makefile (hunk 1 above)
Comment (heschlie) [inline Makefile]: oh nice, I like that

**Comment by heschlie** [comment]
@tomdee I removed that marker, there was no need for it anymore, and am directly untaring into dist, much cleaner.

**Comment by tomdee** [comment]
Cool, please squash then I can merge.

**Comment by heschlie** [comment]
OK @tomdee they are squashed... in a manner of speaking... thanks for the help, again :)

**Pull request merged**